app/src/main/java/ch/epfl/sweng/SDP/shop/ShopActivity.java

```diff
@@ -0,0 +1,330 @@
+
+    private Shop shop;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        overridePendingTransition(0, 0);
+
+        setContentView(R.layout.activity_shop);
+
+        if (enableAnimations) {
+            Glide.with(this).load(R.drawable.background_animation)
+                    .into((ImageView) findViewById(R.id.shopBackgroundAnimation));
+        }
+
+        buyDialog = new Dialog(this);
+        confirmationDialog = new Dialog(this);
+
+        shopColorsRef = Database.getReference("shop.colors");
+
+        typeMuro = Typeface.createFromAsset(getAssets(), "fonts/Muro.otf");
+        typeOptimus = Typeface.createFromAsset(getAssets(), "fonts/Optimus.otf");
+
+        getColorsFromDatabase(shopColorsRef);
+
+        shopItems = findViewById(R.id.shopItems);
+
+        ((TextView) findViewById(R.id.shopMessages)).setTypeface(typeOptimus);
+        ((TextView) findViewById(R.id.yourStars)).setTypeface(typeMuro);
+        ((TextView) findViewById(R.id.yourStars)).setText(String.format(Locale.getDefault(),
+                "%d", Account.getInstance(this).getStars()));
+
+    }
+
+    /**
+     * Accesses the database, gets all the available colors and creates a button in the ScrollView
+     * for each item found.
+     * @param shopColorsReference  Reference to where colors are stored in shop.
+     */
+    protected void getColorsFromDatabase(DatabaseReference shopColorsReference) {
+        shopColorsReference.addListenerForSingleValueEvent(new ValueEventListener() {
+            @Override
+            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
+                if(!isTesting) {
+                    extractColorsFromDataSnapshot(dataSnapshot, Account
+                                                            .getInstance(getApplicationContext()));
+                } else {
+                    fillRandomShop();
+                }
+
+                addColorsToShop();
+            }
+
+            @Override
+            public void onCancelled(@NonNull DatabaseError databaseError) {
+                throw databaseError.toException();
+            }
+        });
+    }
+
+    /**
+     * Tries to extract colors from a snapshot and creates a Button for each.
+     * @param dataSnapshot DataSnapshot from which colors should be extracted.
+     */
+    protected void extractColorsFromDataSnapshot(DataSnapshot dataSnapshot, Account account) {
+        if (dataSnapshot == null) {
+            throw new NullPointerException();
+        }
+
+        shop = new Shop();
+        List<ShopItem> myItems = new LinkedList<>();
+
+        if(Account.getInstance(this).getItemsBought() != null) {
+            myItems = account.getItemsBought();
+        }
+
+        for (DataSnapshot ds : dataSnapshot.getChildren()) {
+            boolean owned = false;
+
+            if (myItems.contains(new ShopItem(ds.getKey(), ds.getValue(int.class)))) {
```

Comment:
You can use a variable for ds.getKey() and ds.getValue(int.class) instead of calling it twice